teps/0122-reproducibility-of-complete-build-instructions.md

```diff
@@ -0,0 +1,698 @@
+---
+status: propoased
+title: Reproducibility of Complete Build Instructions for TaskRuns
+creation-date: '2022-09-14'
+last-updated: '2022-09-14'
+authors:
+- '@chitrangpatel'
+see-also:
+---
+
+# TEP-0122: Reproducibility of Complete Build Instructions for TaskRuns
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+- [Out-of-scope](#out-of-scope)
+- [Future Work](#future-work)
+<!-- /toc -->
+
+## Summary
+This proposal outlines what information is required to reproduce complete build instructions for taskruns. It also suggests where in the provenance to store this information. 
+ 
+## Motivation
+
+[Tekton Chains will report provenance as reproducible if a specific annotation is included in the TaskRun](https://docs.google.com/document/d/12yYP0M_zz5Tc9ftGnLwhDZ_tdZBq4mEod70ngWtHmLs/edit#bookmark=id.1sm6isih1d0n) and it includes Task steps and parameter values in the provenance it generates, however there is a great deal of Task spec contents it ignores (e.g. sidecar definitions) and runtime information it does not include (e.g. workspaces provided at runtime). 
```

Comment:
> I'm inclined to treat missing data like a bug - I'm actually not sure why we aren't just including the whole TaskSpec (@priyawadhwa probably knows more). 🤔
> 
I originally thought of the same thing. I then figured that we probably have more information in the TaskSpec than we need and going through this process might reveal what we actually need. I do agree that including the entire TaskSpec simplifies things a lot !! It will also address some concerns like updating tekton chains to extract new information everything we update (i.e. add/remove fields) the TaskSpec. 

> The `environment` section would be a good target for additional runtime information like workspaces, service accounts, etc.
>
Sounds good!

> To me adding missing data isn't something we need to go through the whole TEP process for - feel free to open an issue in the chains repo if you want to short-circuit the TEP process!

That's a good suggestion.

Reply:
I guess, if we are to store the entire TaskSpec, we should probably store the resolved TaskSpec (i.e. from TaskRun.Status)

Reply:
Just pinging @priyawadhwa for comment here 🙂: Is there a reason why we aren't including the whole TaskSpec in the provenance?

Reply:
Summarizing the discussion from Tekton Chains WG meeting: 
Why don't we just include the entire task/pipeline spec in the provenance?
  -  SLSA does expect certain fields in certain places.
  -  might be tough for policy controllers later if fields are not in the expected spot
Consensus:
  we could start a mapping and then we can always add the full spec later on if it’s a problem.

Reply:
@priyawadhwa @wlynch, one thing I just realized is that if you see the TaskSpec and the underlying Step information below, all the fields except for the `description` of the TaskSpec are `required` for reproducing the build and so it might just be better to include the entire spec and not implement things on the chains side where we cherry pick information from the task spec. If you think that there are additional fields that may not be required, please let me know .